lib/paper_trail/events/base.rb

```diff
@@ -60,15 +60,28 @@ def attribute_changed_in_latest_version?(attr_name)
 
       # @api private
       def nonskipped_attributes_before_change(is_touch)
-        record_attributes = @record.attributes.except(*@record.paper_trail_options[:skip])
+        cache_changed_attributes do
+          record_attributes = @record.attributes.except(*@record.paper_trail_options[:skip])
 
-        Hash[record_attributes.map do |k, v|
-          if @record.class.column_names.include?(k)
-            [k, attribute_in_previous_version(k, is_touch)]
-          else
-            [k, v]
+          record_attributes.each_key do |k|
+            if @record.class.column_names.include?(k)
+              record_attributes[k] = attribute_in_previous_version(k, is_touch)
+            end
           end
-        end]
+        end
+      end
+
+      # Rails 5.1 changed the API of `ActiveRecord::Dirty`.
+      # @api private
+      def cache_changed_attributes
+        if RAILS_GTE_5_1
+          # Everything works fine as it is
+          yield
+        else
+          # Any particular call to `changed_attributes` produces the huge memory allocation.
+          # Lets use the generic AR workaround for that.
+          @record.send(:cache_changed_attributes) { yield }
+        end
```

Comment:
I just seen the strong PT support even for `AR v4.2`, fell that it important to keep it..

As for that particular `if-else` switch around `RAILS_GTE_5_1` - it looks not much tricky than all other alike switches within `Events::Base` class. Even more, if we will remove this switch from `Events::Base`, we will be forced to add it to performance tests, since the memory allocation will be noticeably larger then.

Probably the collapsing of all the `RAILS_GTE_5_1` switches can be a target for `PT v11`.

PS. And, honestly, I need that switch personally, please don't ask me why :)

Reply:
> As for that particular if-else switch around RAILS_GTE_5_1 - it looks not much tricky than all other alike switches ..

Yeah, it's not too painful. We can add it, but ..

> I just seen the strong PT support even for AR v4.2, fell that it important to keep it..

PT follows the rails EoL policy. As soon as rails 6 is released, PT will drop support for rails < 5.2.

> Probably the collapsing of all the RAILS_GTE_5_1 switches can be a target for PT v11.

Yes. When we drop support for old rails versions, SemVer does not require us to make a major version bump, but yes, we often do anyway.

Reply:
> Yeah, it's not too painful. We can add it, but ..
> PT follows the rails EoL policy. As soon as rails 6 is released, PT will drop support for rails < 5.2.

Thanks, I'm ok with that )